### ISSUE TITLE
`tf.dynamic_stitch` gradient is incorrect

### ISSUE THREAD

**Issue by drasmuss** [tensorflow/tensorflow#7397]
### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)
``` python
import tensorflow as tf

x = tf.zeros((1, 3))
y = tf.dynamic_stitch([[0], [0]], [x, tf.ones((1, 3))])

with tf.Session() as sess:
    print("y")
    print(sess.run(y))

    analytic, numeric = tf.test.compute_gradient(x, (1, 3), y, (1, 3))
    print("analytic")
    print(analytic)
    print("numeric")
    print(numeric)
```

gives output
```
y
[[ 1.  1.  1.]]
analytic
[[ 1.  0.  0.]
 [ 0.  1.  0.]
 [ 0.  0.  1.]]
numeric
[[ 0.  0.  0.]
 [ 0.  0.  0.]
 [ 0.  0.  0.]]
```

The numeric gradient correctly shows that `x` has no impact on `y` (since the value of `x` is completely overwritten by a constant in the `dynamic_stitch`).  The analytic gradient is incorrect; it seems like the gradient calculation in `dynamic_stitch` does not handle the case where there are duplicate indices being merged.

**Comment by girving** [comment]
Ug.  You're correct that the gradients are wrong, but I don't see how to fix it without a dramatic performance hit.  Do you have any suggestions?

**Comment by aselle** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by drasmuss** [comment]
The bug still exists, meaning that the `tf.dynamic_stitch` gradients are incorrect.  Is there any other information I can provide that would be helpful?

**Comment by girving** [comment]
Let's leave this open.  Anyone interested should refer to the comments in https://github.com/tensorflow/tensorflow/pull/7487.  The next step would have been to add a new C++ kernel to speed up the bookkeeping required by accurate gradients.

**Comment by bhack** [comment]
Can we close this?

**Comment by drasmuss** [comment]
The gradient implementation is still incorrect, as of TF 2.5.0rc.  Here is an updated example showing the same error
``` python
import tensorflow as tf

x = tf.zeros((1, 3))

analytic, numeric = tf.test.compute_gradient(
    lambda x: tf.dynamic_stitch([[0], [0]], [x, tf.ones((1, 3))]), [x]
)
print("analytic")
print(analytic)
print("numeric")
print(numeric)
```
gives
```
analytic
(array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]], dtype=float32),)
numeric
(array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]], dtype=float32),)
```

**Comment by chunduriv** [comment]
I was able to replicate issues in `tf-nightly 2.10.0-dev20220714`. Please find the [gist](https://colab.sandbox.google.com/gist/chunduriv/27599e19aa2f02b837d459e31da0a335/7397.ipynb#scrollTo=JJ8cPGeX_nVx) for reference. Thank you

**Comment by mohantym** [comment]
Hi @drasmuss !
Just trying to put my observation based on yesterday's trials.
Actually @drasmuss  has used same indices [0] , [0] to stitch to two tensors ([0,0,0] and [1,1,1]. So while trying to allocate tensor at same index , it is taking the max of two tensors and failing the dynamic stitch.

If we use two different indices like [0] , [1] or [0],[2] then results for analytical and theoritical from test.comput_gradient is same.  
@yongtang @bhack 
May be we can put assertion condition in code itself  to check whether user is putting different indices or not .

Attached [gist](https://colab.sandbox.google.com/gist/mohantym/bfecc4cea252cb58a6a23a250d76c514/git_7397.ipynb) for reference.

Thank you!

**Comment by drasmuss** [comment]
Yes, this bug is caused by having duplicate indices. But that is defined and supported behaviour for dynamic stitch (e.g., see [the documentation](https://www.tensorflow.org/api_docs/python/tf/dynamic_stitch)):
> Values are merged in order, so if an index appears in both indices[m][i] and indices[n][j] for (m,i) < (n,j) the slice data[n][j] will appear in the merged result

So if duplicate indices are supported, then the gradient should be correct when using them.

**Comment by mohantym** [comment]
Ok @drasmuss !
Thanks for the update.

**Comment by pjpratik** [comment]
I was able to reproduce this issue in TF Nighly 2.12.0-dev20221218. Please find the gist [here](https://colab.research.google.com/gist/pjpratik/ca25980e1fc5f5e44a486f9efa16a084/7397.ipynb#scrollTo=F2yy848EC0HR). Thank you.

**Comment by synandi** [comment]
I was able to replicate this issue in TF Nighly 2.13.0-dev20230419
. Please find the gist [here](https://colab.sandbox.google.com/gist/synandi/de03fa96504a985ab5006b4497124972/7397.ipynb). Thank you.

**Comment by sushreebarsa** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow version with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the release notes to stay up to date with the latest developments which are happening in the Tensorflow space.

Thank you!

**Comment by drasmuss** [comment]
This bug is still present in TensorFlow 2.16.1. The code from https://github.com/tensorflow/tensorflow/issues/7397#issuecomment-840738297 is still valid, and reproduces the bug. I have edited that into the original post for clarity.

**Comment by Venkat6871** [comment]
Hi @drasmuss ,
Thanks for your patience.
A fix for this issue has been merged in PR #121130 (commit `4554ee6`) and is now available in `tf-nightly`.

Could you please install `tf-nightly` and verify that the gradient behavior is now correct when duplicate indices are stitched? I tested the fix on my end, and it works as expected. I have also attached a [gist](https://colab.sandbox.google.com/gist/Venkat6871/2884fa63895a907b9952feec203c496c/7397_tf_nightly-v.ipynb) for reference. If you can confirm the fix, we can close this issue.

Thank you for bringing this to our attention.
